File: test/Authzi.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraExternalID.NET8.0/Initialization/Starter.fs
namespace Initialization

open Authzi.MicrosoftOrleans.MicrosoftEntra
open Authzi.Security.Authorization
open Authzi.Deploy.MicrosoftEntra.Configuration
open Authzi.MicrosoftEntra
open Authzi.Security
open Authzi.Security.Authorization
open Authzi.Tests.MicrosoftOrleans.MicrosoftEntra
open Authzi.Tests.MicrosoftOrleans.Grains
open Authzi.Tests.MicrosoftOrleans.Grains.SimpleAuthorization
open Authzi.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common
open Microsoft.Extensions.DependencyInjection
open Orleans
open RootConfiguration
open System
open System.Text.Json
open Xunit.Abstractions
open Xunit.Sdk
open Authzi.Identity.MicrosoftEntra
open Microsoft.Identity.Client

[<assembly: Orleans.ApplicationPartAttribute("Authzi.Tests.MicrosoftOrleans.Grains")>]
()

type Starter(messageSink: IMessageSink) =
    inherit XunitTestFramework(messageSink)
    do 
        // Read the configuration.
        let mutable microsoftEntraIdCredentialsJson = Environment.GetEnvironmentVariable("microsoftEntraExternalIdCredentials")

        if String.IsNullOrWhiteSpace(microsoftEntraIdCredentialsJson) then
            microsoftEntraIdCredentialsJson <- Literals.microsoftEntraExternalIDCredentialsJson

        let credentials = JsonSerializer.Deserialize<MicrosoftEntraCredentials>(microsoftEntraIdCredentialsJson)

        // Initialize the test data.
        let entraApp =
            MicrosoftEntraExternalIDApp(credentials.DirectoryId, credentials.WebClient1.Id, 
                credentials.WebClient1.Secret, credentials.WebClient1.AllowedScopes, AadAuthorityAudience.AzureAdMyOrg)

        TestData.UserWithScopeAdeleV <- [[|credentials.AdeleV.Name; credentials.AdeleV.Password; ["Api1"; "Orleans"]|]]
        TestData.UserWithScopeAlexW <- [[|credentials.AlexW.Name; credentials.AlexW.Password; ["Api1"; "Orleans"]|]]
        TestData.Users <- [[| credentials.AdeleV.Name; credentials.AdeleV.Password |]]
        TestData.AzureActiveDirectoryApp <- entraApp
        TestData.Web1Client <- credentials.WebClient1

        TestData.GetAccessTokenForUserOnWebClient1Async <- 
            AccessTokenRetriever.getTokenByUserNameAndPasswordForEntraExternalIDTenant

        let configureSiloHost = fun (services: IServiceCollection) ->
            // Add Azure Active Directory authorization.
            services.AddOrleansAuthorization(TestData.AzureActiveDirectoryApp, 
                fun (config: Authzi.Security.Configuration) -> 
                    config.ConfigureAuthorizationOptions <- Action<AuthorizationOptions>(AuthorizationConfig.ConfigureOptions)
                    ignore())
            // Some custom authorization services.
            AuthorizationConfig.ConfigureServices(services)
        
        // Initialize Orleans test cluster.
        let siloHost = SiloHostBuilder.Build(configureSiloHost)
        siloHost.StartAsync().Wait()

        // Initialize the SiloClient.
        let configureSiloClient = fun (services: IServiceCollection) ->
            let configureCluster (config: Configuration) = 
                                    config.ConfigureAuthorizationOptions <- 
                                        Action<AuthorizationOptions>(AuthorizationConfig.ConfigureOptions)
            AuthorizationConfig.ConfigureServices(services)
            services.AddSingleton<IAccessTokenProvider>( fun _ -> accessTokenProvider) |> ignore
            // Add Azure Active Directory authorization.
            services.AddOrleansClientAuthorization(TestData.AzureActiveDirectoryApp,
                fun config -> configureCluster(config)) |> ignore

        let siloClientHost = SiloClientBuilder.Build(configureSiloClient)
        siloClientHost.StartAsync().Wait()
        TestData.IClusterClient <- siloClientHost.Services.GetService<IClusterClient>()

        //siloClientProvider.SiloClient <- SiloClient.getClusterClient()

module CurrentAssembly =
    [<Literal>]
    let TypeName = "Initialization.Starter"
    [<Literal>]
    let Name = "Authzi.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraExternalID.NET8.0"

[<assembly: Xunit.TestFramework(CurrentAssembly.TypeName, CurrentAssembly.Name)>]
()

File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraExternalID.CSharp/Starter.cs
using AuthZI.Deploy.MicrosoftEntra.Configuration;
using AuthZI.Identity.MicrosoftEntra;
using AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraExternalID.CSharp;
using AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common.Initialization;
using Microsoft.Identity.Client;
using Orleans;
using System;
using System.Text.Json;
using Xunit;
using Xunit.Sdk;
using Xunit.v3;

[assembly: ApplicationPart("AuthZI.Tests.MicrosoftOrleans.Grains")]
[assembly: AssemblyFixture(typeof(ExternalIdMainTestFixture))]
[assembly: Parallelization(Mode = ParallelMode.None)]

namespace AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraExternalID.CSharp;

public class ExternalIdMainTestFixture : MainTestFixture
{
  public ExternalIdMainTestFixture()
  {
    var microsoftEntraExternalIdCredentialsJson =
      Environment.GetEnvironmentVariable("microsoftEntraExternalIdCredentials");

    if (string.IsNullOrWhiteSpace(microsoftEntraExternalIdCredentialsJson))
    {
      microsoftEntraExternalIdCredentialsJson = Literals.microsoftEntraExternalIDCredentialsJson;
    }

    var credentials = JsonSerializer.Deserialize<MicrosoftEntraCredentials>(microsoftEntraExternalIdCredentialsJson)!;
    Credentials = credentials;
    AccessTokenRetriever = new AccessTokenRetriever(true);
  }

  protected override MicrosoftEntraApp CreateMicrosoftEntraApp(
    MicrosoftEntraCredentials credentials,
    Client client) =>
    new MicrosoftEntraExternalIDApp(
      credentials.DirectoryId,
      client.Id,
      client.Secret,
      client.AllowedScopes,
      AadAuthorityAudience.AzureAdMyOrg,
      credentials.Api1.Id);
}
